### ISSUE TITLE
Allow optional output parameters by specifying a default value

### ISSUE THREAD

**Issue by gaoning777** [argoproj/argo#954]
BUG REPORT
When the output resource is not generated in the container, the whole workflow failed.
For example:
```
outputs:
      parameters:
      - name: hello
        valueFrom:
          path: /hello.json
```
When hello.json is not output in the container, the whole job is failed.
Is there anyway to tell argo to continue the workflow/output an empty string when the resource is not found. I believe the output artifact simply generates an empty tar ball when no resource is found.

**Environment**:
- Argo version:
v2.1.1
- Kubernetes version :
Client: 1.11
Server: 1.9+

**Other debugging information (if applicable)**:
- workflow result from `argo get`:
failed to save outputs: exit status 2

**Comment by gaoning777** [comment]
@yebrahim

**Comment by jessesuen** [comment]
This is essentially the same issue as #897. Will duplicate and target this for v2.2

**Comment by jessesuen** [comment]
> When hello.json is not output in the container, the whole job is failed.
> Is there anyway to tell argo to continue the workflow/output an empty string when the resource is > not found. I believe the output artifact simply generates an empty tar ball when no resource is found.

Actually reading it closer, @gaoning777  i think we are on opposite ends of what we feel is the right behavior.

> I believe the output artifact simply generates an empty tar ball when no resource is found.

This is not ideal. For artifacts, the desired behavior is that if someone specifies an artifact path which doesn't exist, then it is considered a user error. Bug #897 is about the fact that we continue down the workflow even though the artifact path does not exist.

For parameters, I think this is also the right behavior, which according to your observations, is already happening for parameters.

One thing we might consider is having default values for output parameters. So if you have the following:
```
outputs:
  parameters:
  - name: hello
    default: ''
    valueFrom:
      path: /hello.json
```

if /hello.json does not exist, then we would return empty string, which I think addresses your use case.

**Comment by gaoning777** [comment]
In terms of workflow continuing when the artifact path is not found, I think it is reasonable to continue the workflow because the correctness of the whole workflow might not depend on the existence of a certain artifact. Users might simply want to output an artifact for some other purposes. Take a step back, it might make sense to make it configurable whether users want the workflow to stop if no artifacts are found.

**Comment by vicaire** [comment]
jessesuen@, yes I think that would solve the problem. Thanks!

**Comment by jessesuen** [comment]
@gaoning777 reopening because if we allow default values for outputs, then it should satisfy yours and @vicaire's request. I think this is a reasonable request.

**Comment by gaoning777** [comment]
Oh, right, this is to-be-added feature.
I thought the default is already supported in output parameters.
Thanks jessesuen

**Comment by jessesuen** [comment]
This should be now doable with PR: https://github.com/argoproj/argo/pull/1277

**Comment by samath117** [comment]
Are default values for output parameters implemented? It's not working for me in 2.3.0. I couldn't find any documentation in that pull request, either.

Minimal non-working example:
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: minimal-default-parameter-example-
spec:
  entrypoint: plan
  templates:
    - name: plan
      container:
        image: alpine:latest
        command: [sh, -c]
        args: ['echo hello']
      outputs:
        parameters:
        - name: message
          valueFrom:
            path: /test-file.txt
          default: "hello"

```

**Comment by cccdy** [comment]
hi，@jessesuen i also have a similar problem， i want to pass some failed log messages to global output， so that other container can access the log messages. i follow your advice for default value，but it still fails with “failed to save outputs: exit status 2”. the yaml snippet  is as follows，waiting for your help，thanks！
    outputs:
      parameters:
      - name: error-msg
        default: ""
        valueFrom:
          path: /tmp/error.txt
        globalName: global-error
